x-pack/filebeat/input/http_endpoint/input.go

```diff
@@ -0,0 +1,279 @@
+	}
+
+	// Create an instance of the HTTP server with the beat context
+	in.server, err = createServer(in)
+	if err != nil {
+		return nil, err
+	}
+
+	in.log.Infof("Initialized %v input on %v:%v", inputName, in.config.ListenAddress, in.config.ListenPort)
+
+	return in, nil
+}
+
+// Run starts the input worker then returns. Only the first invocation
+// will ever start the worker.
+func (in *HttpEndpoint) Run() {
+	in.workerOnce.Do(func() {
+		in.workerWg.Add(1)
+		go in.run()
+	})
+}
+
+func (in *HttpEndpoint) run() {
+	defer in.workerWg.Done()
+	defer in.log.Infof("%v worker has stopped.", inputName)
+	in.server.Start()
+
+	return
+}
+
+// Stops HTTP input and waits for it to finish
+func (in *HttpEndpoint) Stop() {
+	in.workerCancel()
+	in.workerWg.Wait()
+}
+
+// Wait is an alias for Stop.
+func (in *HttpEndpoint) Wait() {
+	in.Stop()
+}
+
+// If middleware validation successed, event is sent
+func (in *HttpEndpoint) sendEvent(w http.ResponseWriter, r *http.Request) {
+	event := in.outlet.OnEvent(beat.Event{
+		Timestamp: time.Now().UTC(),
+		Fields: common.MapStr{
+			in.config.Prefix: in.eventObject,
+		},
+	})
+	if !event {
+		in.sendResponse(w, http.StatusInternalServerError, in.createErrorMessage("Unable to send event"))
+		return
+	}
+
+	return
+}
+
+func (in *HttpEndpoint) sendResponse(w http.ResponseWriter, h uint, b string) {
+	w.WriteHeader(int(h))
+	w.Write([]byte(b))
+}
+
+// Triggers if middleware validation returns successful
+func (in *HttpEndpoint) apiResponse(w http.ResponseWriter, r *http.Request) {
+	objmap := make(map[string]string)
+	json.Unmarshal([]byte(in.config.ResponseHeader), &objmap)
+	for k, v := range objmap {
+		w.Header().Set(k, v)
+	}
+	in.sendEvent(w, r)
+	in.sendResponse(w, uint(in.config.ResponseCode), in.config.ResponseBody)
+}
+
+// Runs all validations for each request
+func (in *HttpEndpoint) validateRequest(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if in.config.BasicAuth {
+			status, err := in.validateAuth(w, r)
+			if err != "" && status != 0 {
+				in.sendResponse(w, status, err)
+				return
+			}
+		}
+
+		status, err := in.validateMethod(w, r)
+		if err != "" && status != 0 {
+			in.sendResponse(w, status, err)
+			return
+		}
+
+		status, err = in.validateHeader(w, r)
+		if err != "" && status != 0 {
+			in.sendResponse(w, status, err)
+			return
+		}
+
+		status, err = in.validateBody(w, r)
+		if err != "" && status != 0 {
+			in.sendResponse(w, status, err)
+			return
+		}
+
+		next.ServeHTTP(w, r)
+	})
+}
+
+// Validate that only supported Accept and Content type headers are used
+func (in *HttpEndpoint) validateHeader(w http.ResponseWriter, r *http.Request) (uint, string) {
+	if r.Header.Get("Content-Type") != "application/json" {
+		return http.StatusUnsupportedMediaType, in.createErrorMessage("Wrong Content-Type header, expecting application/json")
+	}
+
+	if r.Header.Get("Accept") != "application/json" {
```

Comment:
So, my previous comment about overriding headers mainly stems from this line. The `Accept` header checking seems to indicate that we're always going to respond with a `Content-Type: application/json` header--but the fact that we allow the response headers to be overridden in the config actually breaks this. Nothing currently prevents the user from doing something like:

```
{beatname_lc}.inputs:
- type: http_endpoint
  ...
  response_body: '<message>success</message>'
  response_header: '{"Content-Type": "application/xml"}'
```

If we're actually wanting to make sure we always hand back `json` I'd do two things:

1. Always set the `Content-Type: application/json` header and ignore what the user may specify in the config
2. Validate the response value that the user specifies in `response_body` is actually valid JSON.